utils/yoke/src/yokeable.rs

```diff
@@ -0,0 +1,248 @@
+///     }
+///
+///     unsafe fn make(from: Bar<'a>) -> Self {
+///         // We're just doing mem::transmute() here, however Rust is
+///         // not smart enough to realize that Bar<'a> and Bar<'static> are of
+///         // the same size, so instead we use transmute_copy
+///
+///         // This assert will be optimized out, but is included for additional
+///         // peace of mind as we are using transmute_copy
+///         debug_assert!(mem::size_of::<Bar<'a>>() == mem::size_of::<Self>());
+///         let ret = mem::transmute_copy(&from);
+///         mem::forget(from);
+///         ret
+///     }
+///
+///     fn with_mut<F>(&'a mut self, f: F)
+///     where
+///         F: 'static + FnOnce(&'a mut Self::Output),
+///     {
+///         unsafe { f(mem::transmute::<&mut Self, &mut Self::Output>(self)) }
+///     }
+/// }
+/// ```
+pub unsafe trait Yokeable<'a>: 'static {
+    /// This type MUST be `Self` with the `'static` replaced with `'a`, i.e. `Self<'a>`
+    type Output: 'a + Sized;
+
+    /// This method must cast `self` between `&'a Self<'static>` and `&'a Self<'a>`.
+    ///
+    /// # Implementation safety
+    ///
+    /// If the invariants of [`Yokeable`] are being satisfied, the body of this method
+    /// should simply be `{ self }`, though it's acceptable to include additional assertions
+    /// if desired.
+    fn transform(&'a self) -> &'a Self::Output;
+
+    /// This method can be used to cast away `Self<'a>`'s lifetime.
+    ///
+    /// # Safety
+    ///
+    /// The returned value must be destroyed before the data `from` was borrowing from is.
+    ///
+    /// # Implementation safety
+    ///
+    /// A safe implementation of this method must be equivalent to a transmute between
+    /// `Self<'a>` and `Self<'static>`
+    unsafe fn make(from: Self::Output) -> Self;
+
+    /// This method must cast `self` between `&'a mut Self<'static>` and `&'a mut Self<'a>`,
+    /// and pass it to `f`.
+    ///
+    /// # Implementation safety
+    ///
+    /// A safe implementation of this method must be equivalent to a pointer cast/transmute between
+    /// `&mut Self<'a>` and `&mut Self<'static>` being passed to `f`
+    ///
+    /// # Why is this safe?
+    ///
+    /// Typically covariant lifetimes become invariant when hidden behind an `&mut`,
+    /// which is why the implementation of this method cannot just be `f(self)`.
+    /// The reason behind this is that while _reading_ a covariant lifetime that has been cast to a shorter
+    /// one is always safe (this is roughly the definition of a covariant lifetime), writing
+    /// may not necessarily be safe since you could write a smaller reference to it. For example,
+    /// the following code is unsound because it manages to stuff a `'a` lifetime into a `Cow<'static>`
+    ///
+    /// ```rust,compile_fail
+    /// # use std::borrow::Cow;
+    /// # use yoke::Yokeable;
+    /// struct Foo {
+    ///     str: String,
+    ///     cow: Cow<'static, str>,
```

Comment:
Personally I think these examples would have been clearer with a `&'static str`. Shrug.

Reply:
It's mostly because I want to motivate the _safe_ version of this, which is mutating it into a `Cow::Owned`